### ISSUE TITLE
Towards Word By Word Attention Model

### ISSUE THREAD

**Issue by SourKream** [fchollet/keras#2480]
I am implementing a word by word attention model as proposed in [this paper](http://arxiv.org/pdf/1509.06664v1.pdf). I had come up with 2 ways of implementing such a model. 
- The first method I though was to implement an unrolled version of the attention model. This unrolled version has multiple Dense layers that are actually supposed to have the same weigths, but fitting directly may result in different weigths. So I use a callback which, on every batch end, sets the weights of all these Dense layers, that are supposed to be constraint to have the same weights, to the average value of the learnt weights of these layers. (As suggested by @andychisholm in issue #362) This method has been working decently well for me and it takes about 1hr per epoch to train. 
- The second method I thought was to create a custom recurrent node, which extends the Recurrent class. I made this class by looking at the class definition for SimpleRNN node and LSTM node. I have included this code below. Since time was my only concern here and not memory, I wrote the code (to the best that i could think of) only for consume_less=='cpu', although i later also tried (again to the best that i could think of) for consume_less=='mem' and it did not really give me a time boost. I am running the code on a GPU. This node takes around 1 day per epoch to fit.

This is a very drastic difference in terms of time taken by the 2 methods, considering that both the methods are essentially performing the same thing, even though method 2 is what i understand to be the correct way to implement it. Any suggestions for improving the running time for method 2 or any comments on making method 1 work better are welcome.

[attentionModel.txt](https://github.com/fchollet/keras/files/233143/attentionModel.txt)
[Gist](https://gist.github.com/SourKream/670750a77dd0249c8c61f74cb374388c) for the above code.

**Comment by braingineer** [comment]
can you provide a gist (gist.github.com) of the model?  newlines were slaughtered in that .txt

Also, can you be more specific about your first implementation?  I don't understand the weight tying you are referring to. 

Basically, you want a soft attention calculation over the first sentence that gets recalculated for each step in the second sentence?  Also, the r_tp1 is the last soft attention, right?  Interesting they didn't make that more obvious by putting it into the model diagram..

**Comment by braingineer** [comment]
also, if you're using theano, it has a profiling function.  It could help you understand why one model is so slow.  You just pass kwargs into the keras compile function:

```python
theano_kwargs = {'profile':True}
model.compile(loss='categorical_crossentropy', optimizer=optimizer, 
                           metrics=['accuracy'], **theano_kwargs)
```

**Comment by SourKream** [comment]
[Here](https://gist.github.com/SourKream/670750a77dd0249c8c61f74cb374388c) is the gist for the attentionModel.py that I wrote. I didnt know about gists, thanks for introducing me to that.

So in my first implementation method, the weight tying i am referring to comes from equations (11), (12) and (13) from the paper. These equations form a sort of a recurrent layer with the weights **W^r**, **w** and **W^t**. (The weights W^y and W^h can be ignored as u can simply pass the input Y and H through TimeDistributedDense Layers to reach this result). Now when i implement this in Keras in an unrolled fashion, i define all the nodes M_1, M_2, ... M_N, alpha_1, alpha_2, ... alpha_N, r_1, r_2, .... r_N manually which results in there being a group of Dense layers that are all actually supposed to be **W^r** and a group that are all supposed to be **w** and a group that are all supposed to be **W^t**. So within such a group, I have to make all the Dense layers of equal weights.

You are correct in saying that we are calculating the soft attention over the first sentence looking at every word in the second sentence and we are doing this recursively so that the calculation of r_t involves r_{t-1}.

**Comment by SourKream** [comment]
I AM using theano as the backend. Thanks for that tip. Will do check it out.

**Comment by braingineer** [comment]
Hi @SourKream 

I think I'm following your code. Let me see if I've deciphered the `step` function in your rnn:

```python
Y_o = K.dot(self.Y, self.W_y) 

h_o = K.dot(x, self.W_h).dimshuffle((0,'x',1))
h_o_repeated = K.repeat_elements(h_o, L, axis=1)

r_o = K.dot(r_tm1, self.U_r).dimshuffle((0,'x',1))
r_o_repeated = K.repeat_elements(r_o,L, axis=1)

M = K.tanh(Y_o + h_o_repeated + r_o_repeated)        

alpha = K.dot(M, self.W)
alpha = K.softmax(alpha[:,:,0]) 
alpha = alpha.dimshuffle((0,'x',1))

rtm1_weighted = K.tanh(K.dot(r_tm1.dimshuffle((0,'x',1)), self.U_t))
Y_attended = T.batched_dot(alpha, self.Y) 
output = (Y_attended+rmt1_weighted)[:,0,:]

return output, [output]
```

At every time step, you are calculating the output as the weighted combination of Y and the premise representation from the last timestep.   the input to the rnn is h_t (x above) and the last hidden rep.  you MLP align the input to M_t, softly attend it, then combine it with a learned weighting of the last rep.  

Some things I am confused about (and I am ignoring the batch dimension in the following comments, but it doesn't change the functionality at all)

1. Why do you expand r_tm1 to 3dims right before you combine it with U_t?  r_t should be a vector and U_t should be a matrix.  The output of W^t * r_tp1 should be a vector as well

2. I've never really gotten great intuitions on batched dot.  Why is Y_attended a batched dot?  The way I would have done it is K.sum(K.expand_dims(alpha,axis=1) * self.Y), axis=-1) 

By the way, here' s a gist of a soft attention implementation I made (anchored at where I calculated the attention scores).  I'm not sure if it's an issue for you, but zero values were an issue for me (masking and such).  

https://gist.github.com/braingineer/27c6f26755794f6544d83dec2dd27bbb#file-softattention-py-L78

**Comment by LiuAndYang** [comment]
Hi, @SourKream , did you get the same accuracy in that paper using your first code? I mean the accuracy 82.3 %

**Comment by SourKream** [comment]
Hey @braingineer

Yeah I think you have broken down the `step` function correctly. To answer you questions,

_1. Why do you expand r_tm1 to 3dims right before you combine it with U_t? r_t should be a vector and U_t should be a matrix. The output of W^t * r_tp1 should be a vector as well_

- I do not expand r_tm1 to 3dims **before** combining it U_t, but I do so **after** I have combined it with U_t. This is because I want to repeat r_o, L times, along axis=1 so that it matches in dimension with Y_o when we take the sum for finding M (Same as we have done for h_o_repeated from h_o). Dot product of U_t and r_tm1 gives us a vector of dim (BatchSize, k) and we expand and repeat it to make it (BatchSize, L, k).

_2. I've never really gotten great intuitions on batched dot. Why is Y_attended a batched dot? The way I would have done it is K.sum(K.expand_dims(alpha,axis=1) * self.Y), axis=-1)_

- The intuition of the batched dot is as follows. So for every sample in our batch, we have a different alpha vector, correct? which is (because we have expanded the dimension) of size (1, L). And then for every sample in our batch, we have a Y matrix which is of size (L, k). Now we want to iterate over all the samples in our batch and compute this alpha * Y product for each sample. That is given aplha (BatchSize, 1, L) and Y (BatchSize, L, k) we want to do,
```
for i in range(BatchSize):
    output[i,:,:] = alpha[i,:,:] * Y[i,:,:])

```
This is basically what K.batched_dot does.

I'll check out your soft attention model soon and get back to you on it. 
I didnt get what you were saying about the Zero values issue. Since the inputs to my model were of varying sizes, I fixed the size of the input and padded all inputs with a "pad_tok" (which is separate from an "unk" token). I initialise the embedding layer for both these tokens to 0 and I think it learns their weights to differentiate between the two.

**Comment by SourKream** [comment]
Hey @LiuAndYang 

- I am not getting the same accuracy as what is mentioned right now. The best accuracy I could get was around 77-78 %. I trying to figure out what is going wrong and tune the hyper parameters. 

- My entire code for this project is on github [here](https://github.com/SourKream/NLP-Project). However as I am still working on it, it is pretty disorganised and not documented and I am not sure how much you will be able to make from it. Sorry about that. I can point you towards the particular file which contains the code for the first method ([here](https://github.com/SourKream/NLP-Project/blob/master/Code/newModel.py)). I have implemented the weight sharing using a callback over [here](https://github.com/SourKream/NLP-Project/blob/master/Code/newModel.py#L226).

- I think the particular error you are referring to is because of a dimension miss-match in the weights you are assigning to a matrix. My best bet would be that you are assigning a (300,300) dimension matrix to the biases of the Dense Layer which are supposed to be a vector of size (300). The `set_weights` function for a  Dense layer takes a list of 2 elements as arguments. The first element is the weight matrix of size (k,k) and the second is the biases of size (k). I think you are passing it a list containing (k,k) and (k,k).
